iothub_client/inc/iothub_device_client.h

```diff
@@ -349,6 +348,88 @@ extern "C"
+
+    /**
+    * @brief    This API performs an Azure Storage PUT BLOCK LIST operation and
+    *           notifies Azure IoT Hub of the upload completion.
+    * @remark   It is part of a set of functions for more granular control over Azure IoT-based blob uploads.
+    *           This function is expected to be used along with:
+    *           `IoTHubDeviceClient_CreateUploadContext`
+    *           `IoTHubDeviceClient_UploadBlockToBlob`
+    *           `IoTHubDeviceClient_DestroyUploadContext`
+    *           For the standard less-granular uploads to blob please use either
+    *           `IoTHubDeviceClient_UploadToBlobAsync` or `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync`.
+    *           If this function fails (due to any HTTP error to either Azure Storage or Azure IoT Hub) it can
+    *           be run again for a discretionary number of times in an attempt to succeed after, for example,
+    *           an internet connectivity disruption is over.  
+    *
+    * @param    uploadContextHandle    The handle created with `IoTHubDeviceClient_CreateUploadContext`.
+    * @param    isSuccess              A boolean value indicating if the call(s) to `IoTHubDeviceClient_UploadBlockToBlob` succeeded.
+    * @param    responseCode           An user-defined code to signal the status of the upload (e.g., 200 for success, or -1 for abort).
+    * @param    responseMessage        An user-defined status message to go along with `responseCode` on the notification to Azure IoT Hub.
+    *
+    * @warning  This is a synchronous/blocking function.
+    *
+    * @return   IOTHUB_CLIENT_OK upon success or an error code upon failure.
+    */
+    MOCKABLE_FUNCTION(, IOTHUB_CLIENT_RESULT, IoTHubDeviceClient_CompleteUploadToBlob, IOTHUB_CLIENT_LL_UPLOADTOBLOB_CONTEXT_HANDLE, uploadContextHandle, bool, isSuccess, int, responseCode, const char*, responseMessage);
+
+    /**
+    * @brief    This API destroy a blob upload context previously created with `IoTHubDeviceClient_CreateUploadContext`.
+    * @remark   It is part of a set of functions for more granular control over Azure IoT-based blob uploads.
+    *           This function is expected to be used along with:
+    *           `IoTHubDeviceClient_CreateUploadContext`
+    *           `IoTHubDeviceClient_UploadBlockToBlob`
+    *           `IoTHubDeviceClient_CompleteUploadToBlob`
+    *           For the standard less-granular uploads to blob please use either
+    *           `IoTHubDeviceClient_UploadToBlobAsync` or `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync`.
+    *
+    * @param    uploadContextHandle     The handle created with `IoTHubDeviceClient_CreateUploadContext`.
+    *
+    * @warning  This is a synchronous/blocking function.
+    *
+    * @return   Nothing.
+    */
+    MOCKABLE_FUNCTION(, void, IoTHubDeviceClient_DestroyUploadContext, IOTHUB_CLIENT_LL_UPLOADTOBLOB_CONTEXT_HANDLE, uploadContextHandle);
```

Comment:
Any ordering issues if I have 2 contexts and free them in reverse order they were created in?

Reply:
No. Upload contexts are completely independent. I don't think we need to document that, I think. Would you suggest any additional documentation for that?